In [1]:
import os
import torch
import numpy as np
import torchvision.transforms as transforms
from PIL import Image
import matplotlib.pyplot as plt

from zsldataset import ZSLDataset
from models import ContinuousMap, ContinuousMapResidual

def dist_matrix(batch1, batch2):
    delta = batch1.unsqueeze(1) - batch2.unsqueeze(0)
    
    dist_matrix = (delta * delta).mean(dim=-1)
    
    return dist_matrix


def mag(u):
    return torch.dot(u, u)


def dist(u, v):
    return torch.dot(u - v, u - v)

## Load Yahoo dataset

Should be converted into ZSL format beforehand.

In [2]:
trainset = ZSLDataset('Data/APY/train', use_predicates=False, use_irevnet=True)
testset = ZSLDataset('Data/APY/test', use_predicates=False, use_irevnet=True)

In [3]:
bsize = 128
n_epochs = 200
num_classes = trainset.classes.shape[0]

dim_semantic = trainset[0]['class_embedding'].shape[0]
dim_visual = trainset[0]['image_embedding'].shape[0]

all_class_embeddings = torch.tensor(np.array(trainset.class_embeddings)).cuda().float()
classes_enum = torch.tensor(np.array(range(num_classes), dtype=np.int64)).cuda()

In [1]:
v_to_s = ContinuousMap(dim_source=dim_visual, dim_dest=dim_semantic, width=512).cuda()
s_to_v = ContinuousMap(dim_source=dim_semantic, dim_dest=dim_visual, width=512).cuda()

# optimizer = torch.optim.Adam(list(v_to_s.parameters()) + list(s_to_v.parameters()),
#                                 lr = 1e-3,
#                                 betas=(0.9, 0.999),
#                                 weight_decay=3e-2)

optimizer = torch.optim.SGD(list(v_to_s.parameters()) + list(s_to_v.parameters()),
                                lr = 3e-4,
                                momentum=.9,
                                weight_decay=3e-2,
                                   nesterov=True)

NameError: name 'ContinuousMap' is not defined

In [5]:
alpha1 = 50 # triplet
alpha2 = 1e-3 # surjection
alpha3 = 1e-3 # l2 loss

In [6]:
mse = torch.nn.MSELoss()
triplet_loss = torch.nn.TripletMarginLoss(margin=1, p=2)
positive_part = torch.nn.ReLU()

In [7]:
trainloader = torch.utils.data.DataLoader(trainset, 
                                             batch_size=bsize, 
                                             shuffle=True, 
                                             num_workers=4, 
                                             pin_memory=True, 
                                             drop_last=True)

testloader = torch.utils.data.DataLoader(testset, 
                                         batch_size=bsize, 
                                         shuffle=True, 
                                         num_workers=4, 
                                         pin_memory=True, 
                                         drop_last=True)

In [8]:
for e in range(n_epochs):
    v_to_s = v_to_s.train()
    s_to_v = s_to_v.train()
    
    for i, sample in enumerate(trainloader):
        optimizer.zero_grad()
        
        batch_visual = sample['image_embedding'].cuda().float()
        batch_semantic = sample['class_embedding'].cuda().float()

        batch_classes = sample['class_id'].cuda()
        
        e_hat = v_to_s(s_to_v(all_class_embeddings))
        delta = (e_hat - all_class_embeddings)
        surjection_loss = (delta * delta).sum(dim=-1).mean()
        
        s_out = v_to_s(batch_visual)
        
        same_class = classes_enum.unsqueeze(0) == batch_classes.unsqueeze(1)
        same_class = same_class.detach()

        l2_dist_loss = (dist_matrix(s_out, all_class_embeddings) * same_class.float()).mean()
        
        l2_loss = (s_out * s_out).sum(dim=-1).mean()

        loss = alpha1 * l2_dist_loss + alpha2 * surjection_loss + alpha3 * l2_loss
        
        loss.backward()
        
#         c_hat = d_matrix.argmin(dim = -1)
#         print((c_hat == batch_classes).float().mean().item())
        
        optimizer.step()
        
        print(loss.item(), end=', ')
        
    if (e+1) % 50 == 0:
        for param_group in optimizer.param_groups:
            param_group['lr'] = param_group['lr'] * 0.7
    
    if (e+1) % 5 == 0:
        print('\n\n- Evaluation on epoch {}'.format(e))
        
        avg_accuracy = 0.
        avg_loss = 0.
        n = 0
        
        v_to_s = v_to_s.eval() 
        s_to_v = s_to_v.eval() 
        
        with torch.no_grad():
            for i, sample in enumerate(testloader):
                n += 1
                
                batch_visual = sample['image_embedding'].cuda().float()
                batch_semantic = sample['class_embedding'].cuda().float()

                batch_classes = sample['class_id'].cuda()

                s_out = v_to_s(batch_visual)

                same_class = classes_enum.unsqueeze(0) == batch_classes.unsqueeze(1)
                same_class = same_class.detach()

                d_matrix = dist_matrix(s_out, all_class_embeddings)
                
                c_hat = d_matrix.argmin(dim = -1)

                closest_negative, _ = (d_matrix + same_class.float() * 1e6).min(dim=-1)
                furthest_positive, _ = (d_matrix * (1 - same_class.float())).max(dim=-1)

                loss = positive_part(furthest_positive - closest_negative + 5)
                loss = alpha1 * furthest_positive.mean()

                avg_loss += loss.item()
                avg_accuracy += (c_hat == batch_classes).float().mean().item()

        avg_accuracy /= n
        avg_loss /= n

        print('Average acc.: {}, Average loss:{}\n\n'.format(avg_accuracy, avg_loss))


0.4915170967578888, 0.49844157695770264, 0.4957919418811798, 0.4919854402542114, 0.4862467050552368, 0.49787646532058716, 0.48720985651016235, 0.49196577072143555, 0.4912277162075043, 0.49459144473075867, 0.49889445304870605, 0.49497780203819275, 0.48615705966949463, 0.48445624113082886, 0.49102166295051575, 0.48734182119369507, 0.4881494641304016, 0.48333361744880676, 0.4875330626964569, 0.4913105368614197, 0.48321622610092163, 0.48620253801345825, 0.4877685606479645, 0.48689937591552734, 0.476678729057312, 0.4897898733615875, 0.48716309666633606, 0.481413334608078, 0.4879518449306488, 0.4816776514053345, 0.48640763759613037, 0.48394572734832764, 0.4838361442089081, 0.48460763692855835, 0.4829810559749603, 0.4840800166130066, 0.4803541898727417, 0.48158764839172363, 0.47057920694351196, 0.4840848445892334, 0.47674912214279175, 0.48887771368026733, 0.47863277792930603, 0.478762149810791, 0.48071280121803284, 0.47183606028556824, 0.47717946767807007, 0.47423338890075684, 0.4761646986007

0.36103877425193787, 0.35178127884864807, 0.3552195727825165, 0.3548869788646698, 0.36034950613975525, 0.3530515730381012, 0.36104816198349, 0.35972288250923157, 0.35968315601348877, 0.355384886264801, 0.36170950531959534, 0.3564722239971161, 0.3558572828769684, 0.35908636450767517, 0.3609420359134674, 0.3588491976261139, 0.35429948568344116, 0.3569100797176361, 0.35644519329071045, 0.35520637035369873, 0.3577401638031006, 0.35521161556243896, 0.35196080803871155, 0.351428359746933, 0.3551003336906433, 0.3570314645767212, 0.3541903495788574, 0.3483569324016571, 0.3527510166168213, 0.3572862148284912, 0.35220709443092346, 0.3485104739665985, 0.35452115535736084, 0.3542783558368683, 0.359988272190094, 0.35424160957336426, 0.3518451154232025, 0.3463098108768463, 0.3498465418815613, 0.35264670848846436, 0.35524633526802063, 0.34890103340148926, 0.3559247553348541, 0.34747225046157837, 0.3516426384449005, 0.3529510796070099, 0.34411340951919556, 0.35206979513168335, 0.34865036606788635, 0.3

0.2931578457355499, 0.2941986918449402, 0.28929588198661804, 0.29245954751968384, 0.29339367151260376, 0.29200825095176697, 0.2910558879375458, 0.29231786727905273, 0.29428890347480774, 0.29305243492126465, 0.28603237867355347, 0.2920326590538025, 0.292283833026886, 0.28720349073410034, 0.2918691039085388, 0.2898818850517273, 0.2904588282108307, 0.288088858127594, 0.28999587893486023, 0.29499542713165283, 0.29184702038764954, 0.291816383600235, 0.29277506470680237, 0.2931694984436035, 0.28488245606422424, 0.2926607131958008, 0.2868139147758484, 0.2935035228729248, 0.2905697524547577, 0.28900450468063354, 0.2887299060821533, 0.29034948348999023, 0.29491570591926575, 0.2926223874092102, 0.29163241386413574, 0.2859703600406647, 0.2821742594242096, 0.289939820766449, 0.28822362422943115, 0.28645458817481995, 0.2850208580493927, 0.2836058437824249, 0.2885497808456421, 0.29041001200675964, 0.286548912525177, 0.2867641746997833, 0.287227600812912, 0.2841838598251343, 0.291183739900589, 0.2917

0.25122666358947754, 0.25258690118789673, 0.2548067569732666, 0.2521599531173706, 0.2511816918849945, 0.24591445922851562, 0.25150376558303833, 0.25213250517845154, 0.24945981800556183, 0.25212129950523376, 0.25128182768821716, 0.2529321610927582, 0.24779337644577026, 0.2568884491920471, 0.2512436807155609, 0.25440624356269836, 0.2523612678050995, 0.25931093096733093, 0.2550913095474243, 0.24812117218971252, 0.2507515847682953, 0.24985137581825256, 0.2505793571472168, 0.2549118995666504, 0.25780197978019714, 0.25564584136009216, 0.2514263093471527, 0.2585103213787079, 0.2517186105251312, 0.24956965446472168, 0.2507646083831787, 0.2576066255569458, 0.25541576743125916, 0.24932681024074554, 0.25275835394859314, 0.24829526245594025, 0.24474383890628815, 0.25089749693870544, 0.2567232549190521, 0.24505403637886047, 0.24988172948360443, 0.24637021124362946, 0.24307161569595337, 0.25418636202812195, 0.25629863142967224, 0.2480309158563614, 0.24963189661502838, 0.250423401594162, 0.2448242902

0.2253517359495163, 0.22814762592315674, 0.2268962562084198, 0.23010452091693878, 0.22704461216926575, 0.23151132464408875, 0.22778774797916412, 0.22822847962379456, 0.2284429520368576, 0.22412586212158203, 0.2291039079427719, 0.22630317509174347, 0.23714709281921387, 0.22834622859954834, 0.23111428320407867, 0.22688515484333038, 0.22591952979564667, 0.22750017046928406, 0.2230866402387619, 0.21918204426765442, 0.22429926693439484, 0.22772926092147827, 0.22793380916118622, 0.22280924022197723, 0.22493353486061096, 0.22710679471492767, 0.23530790209770203, 0.22527484595775604, 0.2245352864265442, 0.22701424360275269, 0.2284642606973648, 0.22047409415245056, 0.22903871536254883, 0.22853446006774902, 0.22336675226688385, 0.22216369211673737, 0.23066933453083038, 0.22356921434402466, 0.22778648138046265, 0.22900183498859406, 0.2290116250514984, 0.22049693763256073, 0.22861339151859283, 0.23068243265151978, 0.2234964221715927, 0.2279115915298462, 0.22196495532989502, 0.2259484976530075, 0.2

0.21188294887542725, 0.21198628842830658, 0.20869392156600952, 0.2152881920337677, 0.21546317636966705, 0.20744788646697998, 0.2073446363210678, 0.211538627743721, 0.20474475622177124, 0.212133526802063, 0.21354947984218597, 0.20966359972953796, 0.2067110240459442, 0.2104986160993576, 0.2094438523054123, 0.2164597362279892, 0.2160506248474121, 0.2045631855726242, 0.2162744700908661, 0.21010221540927887, 0.20924656093120575, 0.20744948089122772, 0.20985111594200134, 0.20789271593093872, 0.21390286087989807, 0.21383434534072876, 0.21357044577598572, 0.20580828189849854, 0.211979478597641, 0.2120688557624817, 0.21349410712718964, 0.2094554603099823, 0.2080412358045578, 0.20723509788513184, 0.2075602263212204, 0.20993061363697052, 0.207101970911026, 0.2052120566368103, 0.20904190838336945, 0.21328186988830566, 0.21704569458961487, 0.20895525813102722, 0.2096550613641739, 0.20671765506267548, 0.212490513920784, 0.21166639029979706, 0.21157152950763702, 0.21823768317699432, 0.210909679532051

0.20315062999725342, 0.20065422356128693, 0.20217835903167725, 0.2062157541513443, 0.2068314552307129, 0.20244723558425903, 0.20855370163917542, 0.19994503259658813, 0.20774231851100922, 0.20524165034294128, 0.2022569328546524, 0.20287498831748962, 0.2017894983291626, 0.20471245050430298, 0.1964080035686493, 0.206532284617424, 0.20443663001060486, 0.1959277242422104, 0.19681110978126526, 0.20024161040782928, 0.19705821573734283, 0.2016770988702774, 0.19495341181755066, 0.20901966094970703, 0.19793973863124847, 0.20259472727775574, 0.20612259209156036, 0.20418499410152435, 0.19607388973236084, 0.20018315315246582, 0.2022397220134735, 0.20320729911327362, 0.20060262084007263, 0.2089410275220871, 0.19623307883739471, 0.2018972486257553, 0.2045612335205078, 0.1997450590133667, 0.20483790338039398, 0.1988949030637741, 0.20194430649280548, 0.2040369063615799, 0.1983727514743805, 0.19669954478740692, 0.20443223416805267, 0.2040301412343979, 0.19892553985118866, 0.20406652987003326, 0.20198185

0.20104095339775085, 0.1939157396554947, 0.19986951351165771, 0.1936798244714737, 0.19843095541000366, 0.19146478176116943, 0.19933569431304932, 0.20115435123443604, 0.19273757934570312, 0.1923285722732544, 0.1930765062570572, 0.19495798647403717, 0.19569434225559235, 0.1952507644891739, 0.2016109824180603, 0.20109444856643677, 0.1911838799715042, 0.19366760551929474, 0.19205637276172638, 0.1956903487443924, 0.1856742799282074, 0.19758479297161102, 0.19443190097808838, 0.19989579916000366, 0.18954302370548248, 0.1993120163679123, 0.19386456906795502, 0.19623368978500366, 0.19183006882667542, 0.1936047077178955, 0.1949789971113205, 0.19459103047847748, 0.1961127668619156, 0.1976977288722992, 0.190872460603714, 0.19143609702587128, 0.19527947902679443, 0.1935582309961319, 0.1962558925151825, 0.19577376544475555, 0.19825802743434906, 0.19109146296977997, 0.1892639845609665, 0.19522790610790253, 0.19495275616645813, 0.19244393706321716, 0.19511249661445618, 0.1981632560491562, 0.1916535496

0.18964973092079163, 0.1863805204629898, 0.19085508584976196, 0.1866389364004135, 0.1874072253704071, 0.1897616684436798, 0.19067828357219696, 0.18823547661304474, 0.18822649121284485, 0.18651878833770752, 0.18715731799602509, 0.19071052968502045, 0.18798072636127472, 0.18878692388534546, 0.19184501469135284, 0.18839935958385468, 0.18479956686496735, 0.18748202919960022, 0.19710740447044373, 0.19252580404281616, 0.18882495164871216, 0.1887793391942978, 0.19071190059185028, 0.18670353293418884, 0.18554182350635529, 0.19175411760807037, 0.18789617717266083, 0.18898025155067444, 0.18977577984333038, 0.18594816327095032, 0.18684148788452148, 0.1929071694612503, 0.18667608499526978, 0.1867789924144745, 0.19207830727100372, 0.18600957095623016, 0.19348694384098053, 0.18584860861301422, 0.18953251838684082, 0.18964865803718567, 0.18969760835170746, 0.19068700075149536, 0.1878230720758438, 0.19003280997276306, 0.19779576361179352, 0.18932950496673584, 0.18873710930347443, 0.18981584906578064, 

0.18525810539722443, 0.18372707068920135, 0.18207645416259766, 0.1833454966545105, 0.1841072142124176, 0.18276701867580414, 0.18405422568321228, 0.18846429884433746, 0.1844448745250702, 0.18502426147460938, 0.1816568523645401, 0.18377496302127838, 0.1871255785226822, 0.18468432128429413, 0.19219139218330383, 0.18464405834674835, 0.18275922536849976, 0.18264958262443542, 0.18418091535568237, 0.18296535313129425, 0.18037988245487213, 0.1797822266817093, 0.18421171605587006, 0.19053716957569122, 0.1773485243320465, 0.18438847362995148, 0.18749429285526276, 0.1897009164094925, 0.18654510378837585, 0.18561211228370667, 0.1870245337486267, 0.18072956800460815, 0.1874973326921463, 0.1829180270433426, 0.1848851442337036, 0.1854579597711563, 0.18115447461605072, 0.18198516964912415, 0.1873198002576828, 0.18568484485149384, 0.18574877083301544, 0.1835460066795349, 0.18535707890987396, 0.1864992380142212, 0.18150442838668823, 0.17906253039836884, 0.18576079607009888, 0.18168680369853973, 0.182099

0.18098491430282593, 0.1764165461063385, 0.17900687456130981, 0.1803533434867859, 0.18306730687618256, 0.18196070194244385, 0.1818113476037979, 0.1808873862028122, 0.17895475029945374, 0.1854066252708435, 0.18053773045539856, 0.178024560213089, 0.18090669810771942, 0.18143834173679352, 0.1775202751159668, 0.18150977790355682, 0.1815313994884491, 0.1760048121213913, 0.17548251152038574, 0.17854562401771545, 0.1807437539100647, 0.1796920895576477, 0.18424896895885468, 0.17906782031059265, 0.17667916417121887, 0.18092915415763855, 0.17637179791927338, 0.1758372038602829, 0.17878611385822296, 0.1840025782585144, 0.18100287020206451, 0.17384149134159088, 0.1802234947681427, 0.18934333324432373, 0.1779550313949585, 0.18260134756565094, 0.18291744589805603, 0.17733925580978394, 0.1827722191810608, 0.17678087949752808, 0.18344444036483765, 0.18272124230861664, 0.1854955404996872, 0.1800837516784668, 0.18258583545684814, 0.18081827461719513, 0.17753231525421143, 0.18166540563106537, 0.179341733

0.1768285036087036, 0.174711212515831, 0.173237144947052, 0.17601896822452545, 0.17043747007846832, 0.1792542189359665, 0.18506918847560883, 0.17311686277389526, 0.18420903384685516, 0.1771845817565918, 0.18224449455738068, 0.1851799488067627, 0.17977477610111237, 0.17916987836360931, 0.17930293083190918, 0.1733110100030899, 0.18123391270637512, 0.1798572540283203, 0.1776503175497055, 0.17488838732242584, 0.17691932618618011, 0.17876887321472168, 0.1763499677181244, 0.17236916720867157, 0.17600961029529572, 0.17726360261440277, 0.18694493174552917, 0.18044337630271912, 0.18036194145679474, 0.17737208306789398, 0.17878803610801697, 0.18122164905071259, 0.17901277542114258, 0.17896494269371033, 0.1810658574104309, 0.17676924169063568, 0.17628604173660278, 0.17889949679374695, 0.1769237220287323, 0.17938141524791718, 0.17958228290081024, 0.17933301627635956, 0.18219321966171265, 0.17201261222362518, 0.1758987009525299, 0.17396308481693268, 0.17655256390571594, 0.17639845609664917, 0.17554

0.176860973238945, 0.1717321276664734, 0.18062664568424225, 0.1705281138420105, 0.1780274212360382, 0.18167254328727722, 0.17418734729290009, 0.17137664556503296, 0.18153679370880127, 0.17292679846286774, 0.18062201142311096, 0.17384737730026245, 0.17634961009025574, 0.17333248257637024, 0.17025208473205566, 0.17828251421451569, 0.17868278920650482, 0.1789606660604477, 0.17439158260822296, 0.17939899861812592, 0.1770007163286209, 0.17121902108192444, 0.178540900349617, 0.17663265764713287, 0.1747511923313141, 0.179545596241951, 0.17965887486934662, 0.17801573872566223, 0.17141273617744446, 0.17144790291786194, 0.17175449430942535, 0.17571519315242767, 0.17641235888004303, 0.18130075931549072, 0.17547646164894104, 0.17686554789543152, 0.17345482110977173, 0.175736203789711, 0.1722334772348404, 0.17404983937740326, 0.17610427737236023, 0.17913192510604858, 0.17306490242481232, 0.17664524912834167, 0.17111632227897644, 0.17652636766433716, 0.175384983420372, 0.17688627541065216, 0.1763503

0.17387360334396362, 0.17325982451438904, 0.17930951714515686, 0.17637504637241364, 0.17289753258228302, 0.16970035433769226, 0.17417874932289124, 0.17276065051555634, 0.17342819273471832, 0.16999982297420502, 0.17608219385147095, 0.17375656962394714, 0.17424127459526062, 0.17434047162532806, 0.1750027984380722, 0.17905478179454803, 0.17257456481456757, 0.16581149399280548, 0.17745189368724823, 0.17802280187606812, 0.17120304703712463, 0.17883481085300446, 0.17335225641727448, 0.17263901233673096, 0.16906635463237762, 0.17423348128795624, 0.16992402076721191, 0.17148445546627045, 0.1781662106513977, 0.17241401970386505, 0.17180851101875305, 0.17255717515945435, 0.17382341623306274, 0.17014741897583008, 0.1700795292854309, 0.1761261522769928, 0.17343690991401672, 0.17128290235996246, 0.17680130898952484, 0.17638376355171204, 0.17614179849624634, 0.177672877907753, 0.17189782857894897, 0.1771055907011032, 0.17337174713611603, 0.17871488630771637, 0.1661520153284073, 0.17236079275608063, 

0.16959066689014435, 0.16637273132801056, 0.1722966581583023, 0.17211802303791046, 0.17053431272506714, 0.1687890887260437, 0.16952648758888245, 0.17036288976669312, 0.17326650023460388, 0.1703985631465912, 0.17560388147830963, 0.17715929448604584, 0.17329005897045135, 0.1673302799463272, 0.17523647844791412, 0.1809362918138504, 0.16916002333164215, 0.16674526035785675, 0.17250601947307587, 0.16797477006912231, 0.1658596247434616, 0.17245995998382568, 0.17064912617206573, 0.16767245531082153, 0.17065967619419098, 0.17479458451271057, 0.1713479608297348, 0.17472349107265472, 0.1724892258644104, 0.1712021678686142, 0.1751604825258255, 0.17608927190303802, 0.17678916454315186, 0.1711290031671524, 0.1753566861152649, 0.17103490233421326, 0.1734277755022049, 0.17104417085647583, 0.17118149995803833, 0.17193454504013062, 0.17167536914348602, 0.16873392462730408, 0.17067663371562958, 0.17434494197368622, 0.17571841180324554, 0.17028743028640747, 0.17426609992980957, 0.17134544253349304, 0.173

0.1730232685804367, 0.16996024549007416, 0.17437432706356049, 0.1632368415594101, 0.17012649774551392, 0.16940470039844513, 0.1627974957227707, 0.16959233582019806, 0.16717864573001862, 0.1718488186597824, 0.17316193878650665, 0.17176972329616547, 0.16946807503700256, 0.16641415655612946, 0.17013512551784515, 0.1677548736333847, 0.17499622702598572, 0.17276141047477722, 0.1679536998271942, 0.17151859402656555, 0.17077386379241943, 0.16560669243335724, 0.17122703790664673, 0.16363216936588287, 0.1710658073425293, 0.1723642796278, 0.1717401146888733, 0.17151041328907013, 0.17283833026885986, 0.17126049101352692, 0.1739082634449005, 0.17228789627552032, 0.16565349698066711, 0.1738520711660385, 0.16839443147182465, 0.1683400571346283, 0.16963407397270203, 0.17813639342784882, 0.16673442721366882, 0.16620957851409912, 0.16460275650024414, 0.1756393313407898, 0.16652584075927734, 0.1692405790090561, 0.1711650937795639, 0.16791851818561554, 0.1706666797399521, 0.1744910329580307, 0.1722055524

0.17020577192306519, 0.17119204998016357, 0.16761058568954468, 0.16714075207710266, 0.17114020884037018, 0.16716428101062775, 0.16892170906066895, 0.171829953789711, 0.1717681586742401, 0.16552157700061798, 0.16707539558410645, 0.16950717568397522, 0.16712386906147003, 0.1695064753293991, 0.17348408699035645, 0.16777586936950684, 0.17029154300689697, 0.1782664954662323, 0.1642875224351883, 0.1636202484369278, 0.1728542298078537, 0.1623554825782776, 0.1708381325006485, 0.171157106757164, 0.1736382246017456, 0.1681414246559143, 0.17209884524345398, 0.17080359160900116, 0.16918668150901794, 0.1684468686580658, 0.16547146439552307, 0.17011065781116486, 0.16874678432941437, 0.16806064546108246, 0.1676001399755478, 0.16926762461662292, 0.17307765781879425, 0.17289452254772186, 0.1684093028306961, 0.17136408388614655, 0.16649633646011353, 0.1684269905090332, 0.16690106689929962, 0.16707175970077515, 0.16901373863220215, 0.1723092645406723, 0.16571420431137085, 0.16969481110572815, 0.170892581

0.1685832291841507, 0.16629518568515778, 0.1697106510400772, 0.16667094826698303, 0.16884036362171173, 0.1692608743906021, 0.1710437536239624, 0.17150311172008514, 0.1678933948278427, 0.16511808335781097, 0.17038722336292267, 0.16659575700759888, 0.16508157551288605, 0.16625013947486877, 0.1656084507703781, 0.16640891134738922, 0.17282892763614655, 0.16464164853096008, 0.16916495561599731, 0.1674347221851349, 0.16745419800281525, 0.1688700020313263, 0.17191103100776672, 0.1662350744009018, 0.16684266924858093, 0.17112624645233154, 0.16622430086135864, 0.17072021961212158, 0.1667444258928299, 0.16591192781925201, 0.17009314894676208, 0.16688060760498047, 0.17449790239334106, 0.172914057970047, 0.1680988371372223, 0.16894254088401794, 0.1687067151069641, 0.16212384402751923, 0.16570794582366943, 0.16797852516174316, 0.16784310340881348, 0.16654974222183228, 0.1744525134563446, 0.16997230052947998, 0.1640445441007614, 0.16137275099754333, 0.166845440864563, 0.17171798646450043, 0.17195616

0.16382364928722382, 0.1690044403076172, 0.16811612248420715, 0.16928815841674805, 0.16778500378131866, 0.1630990207195282, 0.16785532236099243, 0.17326749861240387, 0.16943158209323883, 0.16594882309436798, 0.16211585700511932, 0.17381136119365692, 0.166078120470047, 0.16740094125270844, 0.174429789185524, 0.1657198965549469, 0.16646139323711395, 0.16818910837173462, 0.17087188363075256, 0.16751629114151, 0.16627193987369537, 0.1655140370130539, 0.1668466031551361, 0.1680850237607956, 0.17242948710918427, 0.16968971490859985, 0.16944968700408936, 0.16719530522823334, 0.1657509058713913, 0.16588705778121948, 0.16094040870666504, 0.1700124591588974, 0.16456729173660278, 0.16741937398910522, 0.1635550558567047, 0.16848070919513702, 0.16672208905220032, 0.16816246509552002, 0.16876792907714844, 0.1714102029800415, 0.16990238428115845, 0.16760872304439545, 0.16613450646400452, 0.1626633256673813, 0.1661309003829956, 0.16431571543216705, 0.16697847843170166, 0.17063471674919128, 0.166044011

0.16279667615890503, 0.16688141226768494, 0.1676826775074005, 0.16631406545639038, 0.1660911738872528, 0.16598355770111084, 0.17184868454933167, 0.1705368161201477, 0.16723693907260895, 0.16757433116436005, 0.16403783857822418, 0.1690928339958191, 0.1726686954498291, 0.16620667278766632, 0.16959287226200104, 0.16450360417366028, 0.16515588760375977, 0.17124715447425842, 0.1672040969133377, 0.1636640578508377, 0.16547173261642456, 0.16527514159679413, 0.16954198479652405, 0.1643037497997284, 0.16881385445594788, 0.16842854022979736, 0.16464917361736298, 0.16582369804382324, 0.16773663461208344, 0.1639973223209381, 0.16655518114566803, 0.1658642739057541, 0.16634143888950348, 0.16397175192832947, 0.16466160118579865, 0.16712650656700134, 0.1737411618232727, 0.16845791041851044, 0.167279914021492, 0.16422444581985474, 0.16512872278690338, 0.16526174545288086, 0.16040991246700287, 0.16166454553604126, 0.16321994364261627, 0.16396145522594452, 0.16658876836299896, 0.16496554017066956, 0.167

# Best results

with words only: 34.6%

with words+attributes:

# With attributes

In [1]:
import os
import torch
import numpy as np
import torchvision.transforms as transforms
from PIL import Image
import matplotlib.pyplot as plt

from zsldataset import ZSLDataset
from models import ContinuousMap, ContinuousMapResidual

def dist_matrix(batch1, batch2):
    delta = batch1.unsqueeze(1) - batch2.unsqueeze(0)
    
    dist_matrix = (delta * delta).mean(dim=-1)
    
    return dist_matrix


def mag(u):
    return torch.dot(u, u)


def dist(u, v):
    return torch.dot(u - v, u - v)

In [2]:
from models import EncoderAttributes, DecoderAttributes

In [3]:
trainset = ZSLDataset('Data/APY/train', use_predicates=True, use_irevnet=True)
testset = ZSLDataset('Data/APY/test', use_predicates=True, use_irevnet=True)

In [4]:
bsize = 128
n_epochs = 200
num_classes = trainset.classes.shape[0]

dim_semantic = trainset[0]['class_embedding'].shape[0]
dim_visual = trainset[0]['image_embedding'].shape[0]
dim_attributes = trainset[0]['class_predicates'].shape[0]

all_class_embeddings = torch.tensor(np.array(trainset.class_embeddings)).cuda().float()
all_class_predicates = torch.tensor(np.array(trainset.class_predicates)).cuda().float()
classes_enum = torch.tensor(np.array(range(num_classes), dtype=np.int64)).cuda()

In [5]:
v_to_s = DecoderAttributes(dim_source=dim_visual, dim_target1=dim_attributes, dim_target2=dim_semantic, width=512).cuda()
s_to_v = EncoderAttributes(dim_source1=dim_semantic, dim_source2=dim_attributes, dim_target=dim_visual, width=512).cuda()

# optimizer = torch.optim.Adam(list(v_to_s.parameters()) + list(s_to_v.parameters()),
#                                 lr = 1e-3,
#                                 betas=(0.9, 0.999),
#                                 weight_decay=3e-2)

optimizer = torch.optim.SGD(list(v_to_s.parameters()) + list(s_to_v.parameters()),
                                lr = 1e-4,
                                momentum=.9,
                                weight_decay=3e-3,
                                nesterov=True)

In [6]:
mse = torch.nn.MSELoss()
triplet_loss = torch.nn.TripletMarginLoss(margin=1, p=2)
positive_part = torch.nn.ReLU()

In [7]:
trainloader = torch.utils.data.DataLoader(trainset, 
                                             batch_size=bsize, 
                                             shuffle=True, 
                                             num_workers=4, 
                                             pin_memory=True, 
                                             drop_last=True)

testloader = torch.utils.data.DataLoader(testset, 
                                         batch_size=bsize, 
                                         shuffle=True, 
                                         num_workers=4, 
                                         pin_memory=True, 
                                         drop_last=True)

In [8]:
gamma = 0.4

alpha1 = 10 # triplet
alpha2 = 1e-3 # surjection
alpha3 = 1e-3 # l2 loss


In [9]:
for e in range(n_epochs):
    v_to_s = v_to_s.train()
    s_to_v = s_to_v.train()
    
    for i, sample in enumerate(trainloader):
        optimizer.zero_grad()
        
        batch_visual = sample['image_embedding'].cuda().float()
        batch_semantic = sample['class_embedding'].cuda().float()

        batch_classes = sample['class_id'].cuda()
         
        e_hat = v_to_s(s_to_v(all_class_embeddings, all_class_predicates))
        delta = (e_hat[1] - all_class_embeddings)
        surjection_loss = (delta * delta).sum(dim=-1).mean()
        delta = (e_hat[0] - all_class_predicates)
        surjection_loss = (1-gamma) * surjection_loss + gamma * (delta * delta).sum(dim=-1).mean()
        
        s_out = v_to_s(batch_visual)
        s_attr, s_word = s_out
        
        same_class = classes_enum.unsqueeze(0) == batch_classes.unsqueeze(1)
        same_class = same_class.detach()

        d_matrix = (1 - gamma) * dist_matrix(s_word, all_class_embeddings) + \
                    gamma * dist_matrix(s_attr, all_class_predicates)
        
        ###
        l2_dist_loss = (d_matrix * same_class.float()).mean()
        
        l2_loss = (1-gamma) * (s_word * s_word).sum(dim=-1).mean() + \
                    gamma * (s_attr * s_attr).sum(dim=-1).mean()

        loss = positive_part(furthest_positive - closest_negative + 1)
        loss = alpha1 * loss.mean() + alpha2 * surjection_loss + alpha3 * l2_loss
        
        loss.backward()
        
#         c_hat = d_matrix.argmin(dim = -1)
#         print((c_hat == batch_classes).float().mean().item())
        
        optimizer.step()
        
        print(loss.item(), end=', ')
        
    if (e+1) % 70 == 0:
        for param_group in optimizer.param_groups:
            param_group['lr'] = param_group['lr'] * 0.7
    
    if (e+1) % 5 == 0:
        print('\n\n- Evaluation on epoch {}'.format(e))
        
        avg_accuracy = 0.
        avg_loss = 0.
        n = 0
        
        v_to_s = v_to_s.eval() 
        s_to_v = s_to_v.eval() 
        
        with torch.no_grad():
            for i, sample in enumerate(testloader):
                n += 1
                
                batch_visual = sample['image_embedding'].cuda().float()
                batch_semantic = sample['class_embedding'].cuda().float()

                batch_classes = sample['class_id'].cuda()

                s_out = v_to_s(batch_visual)
                s_attr, s_word = s_out

                same_class = classes_enum.unsqueeze(0) == batch_classes.unsqueeze(1)
                same_class = same_class.detach()

                d_matrix = (1 - gamma) * dist_matrix(s_word, all_class_embeddings) + \
                                gamma * dist_matrix(s_attr, all_class_predicates)
                
                c_hat = d_matrix.argmin(dim = -1)

                closest_negative, _ = (d_matrix + same_class.float() * 1e6).min(dim=-1)
                furthest_positive, _ = (d_matrix * (1 - same_class.float())).max(dim=-1)

                loss = positive_part(furthest_positive - closest_negative + 5)
                loss = alpha1 * furthest_positive.mean()

                avg_loss += loss.item()
                avg_accuracy += (c_hat == batch_classes).float().mean().item()

        avg_accuracy /= n
        avg_loss /= n

        print('Average acc.: {}, Average loss:{}\n\n'.format(avg_accuracy, avg_loss))




10.576552391052246, 10.571242332458496, 10.585657119750977, 10.570113182067871, 10.572693824768066, 10.564399719238281, 10.583442687988281, 10.563982009887695, 10.576849937438965, 10.566195487976074, 10.566062927246094, 10.572005271911621, 10.570043563842773, 10.579184532165527, 10.56896686553955, 10.578373908996582, 10.57557201385498, 10.574932098388672, 10.575716018676758, 10.579255104064941, 10.565694808959961, 10.569279670715332, 10.571930885314941, 10.576912879943848, 10.569809913635254, 10.577178001403809, 10.573882102966309, 10.56824779510498, 10.568777084350586, 10.569540023803711, 10.567770004272461, 10.564340591430664, 10.56863021850586, 10.567337036132812, 10.565375328063965, 10.571569442749023, 10.564568519592285, 10.549568176269531, 10.569628715515137, 10.561238288879395, 10.570425987243652, 10.567354202270508, 10.57348346710205, 10.555363655090332, 10.573796272277832, 10.557287216186523, 10.563777923583984, 10.557394981384277, 10.568008422851562, 10.55777359008789, 10.558

10.473814964294434, 10.472463607788086, 10.471417427062988, 10.47845458984375, 10.471652030944824, 10.477463722229004, 10.476791381835938, 10.465911865234375, 10.474359512329102, 10.47663688659668, 10.476705551147461, 10.47421646118164, 10.475882530212402, 10.474130630493164, 10.475974082946777, 10.466058731079102, 10.470616340637207, 10.476666450500488, 10.47673225402832, 10.468131065368652, 10.47529411315918, 10.473690032958984, 10.471055030822754, 10.477643966674805, 10.472545623779297, 10.476994514465332, 10.471497535705566, 10.475738525390625, 10.475276947021484, 10.477211952209473, 10.472213745117188, 10.477034568786621, 10.467114448547363, 10.470785140991211, 10.467985153198242, 10.469478607177734, 10.46585750579834, 10.468551635742188, 10.46728229522705, 10.474327087402344, 10.469099998474121, 10.47754192352295, 10.475008964538574, 10.472349166870117, 10.467121124267578, 10.469367027282715, 10.469293594360352, 10.469952583312988, 10.474010467529297, 10.463894844055176, 10.47112

10.421088218688965, 10.428439140319824, 10.421414375305176, 10.431675910949707, 10.426183700561523, 10.425251960754395, 10.42251968383789, 10.430203437805176, 10.427746772766113, 10.427931785583496, 10.424474716186523, 10.423203468322754, 10.421954154968262, 10.422307968139648, 10.42325496673584, 10.42492961883545, 10.42291259765625, 10.4251070022583, 10.42514705657959, 10.422247886657715, 10.4263277053833, 10.424420356750488, 10.421612739562988, 10.421426773071289, 10.420987129211426, 10.41860294342041, 10.419031143188477, 10.423933029174805, 10.42479133605957, 10.413914680480957, 10.430465698242188, 10.425638198852539, 10.422989845275879, 10.42755126953125, 10.422779083251953, 10.423589706420898, 10.42320442199707, 10.422731399536133, 10.415956497192383, 10.420620918273926, 10.418737411499023, 10.419844627380371, 10.420681953430176, 10.419843673706055, 10.422870635986328, 10.426002502441406, 10.426702499389648, 10.421138763427734, 10.420893669128418, 10.411648750305176, 10.4199552536

10.389780044555664, 10.388581275939941, 10.392325401306152, 10.385407447814941, 10.385101318359375, 10.387079238891602, 10.39152717590332, 10.383692741394043, 10.385562896728516, 10.387141227722168, 10.392574310302734, 10.3879976272583, 10.389190673828125, 10.384090423583984, 10.38803482055664, 10.386956214904785, 10.381670951843262, 10.386515617370605, 10.384057998657227, 10.385222434997559, 10.391252517700195, 10.39074420928955, 10.388833045959473, 10.38326358795166, 10.388726234436035, 10.382406234741211, 10.386628150939941, 10.3837890625, 10.38699722290039, 10.38428783416748, 10.385071754455566, 10.385074615478516, 10.384552955627441, 10.3930025100708, 10.38403606414795, 10.382936477661133, 10.384008407592773, 10.389664649963379, 10.388554573059082, 10.385189056396484, 10.387979507446289, 10.384016990661621, 10.384309768676758, 10.387726783752441, 10.388214111328125, 10.389104843139648, 10.390302658081055, 10.385220527648926, 10.388657569885254, 10.385799407958984, 10.3816080093383

10.356931686401367, 10.353549003601074, 10.356621742248535, 10.35800552368164, 10.361464500427246, 10.36059284210205, 10.359007835388184, 10.355916023254395, 10.357375144958496, 10.358688354492188, 10.360599517822266, 10.35999870300293, 10.357586860656738, 10.356282234191895, 10.357471466064453, 10.355326652526855, 10.356111526489258, 10.355250358581543, 10.361103057861328, 10.360528945922852, 10.360970497131348, 10.360697746276855, 10.360490798950195, 10.360000610351562, 10.359432220458984, 10.35892391204834, 10.358602523803711, 10.354757308959961, 10.352926254272461, 10.354562759399414, 10.362146377563477, 10.359034538269043, 10.363341331481934, 10.361105918884277, 10.354168891906738, 10.354211807250977, 10.362714767456055, 10.357831001281738, 10.360580444335938, 10.359580993652344, 10.35484790802002, 10.354098320007324, 10.356512069702148, 10.360666275024414, 10.35781478881836, 10.359169960021973, 10.356884956359863, 10.361504554748535, 10.358880043029785, 10.355039596557617, 10.351

10.335402488708496, 10.330821990966797, 10.333220481872559, 10.332305908203125, 10.335033416748047, 10.333381652832031, 10.333087921142578, 10.334651947021484, 10.337987899780273, 10.335518836975098, 10.330207824707031, 10.337191581726074, 10.336637496948242, 10.333478927612305, 10.335897445678711, 10.3360013961792, 10.334919929504395, 10.333841323852539, 10.335259437561035, 10.334000587463379, 10.335041999816895, 10.333409309387207, 10.333932876586914, 10.332616806030273, 10.330013275146484, 10.336058616638184, 10.333253860473633, 10.33336353302002, 10.334792137145996, 10.335355758666992, 10.332635879516602, 10.337862968444824, 10.334906578063965, 10.32772445678711, 10.336648941040039, 10.334996223449707, 10.333101272583008, 10.33915901184082, 10.334662437438965, 10.33344841003418, 10.329483985900879, 10.335226058959961, 10.339251518249512, 10.331744194030762, 10.334752082824707, 10.337077140808105, 10.336529731750488, 10.336921691894531, 10.330901145935059, 10.331003189086914, 10.333

10.314218521118164, 10.310654640197754, 10.314449310302734, 10.31346607208252, 10.310165405273438, 10.312024116516113, 10.31214714050293, 10.312286376953125, 10.315840721130371, 10.312995910644531, 10.31378173828125, 10.30943775177002, 10.313395500183105, 10.316882133483887, 10.313406944274902, 10.313036918640137, 10.315472602844238, 10.316473007202148, 10.314889907836914, 10.315461158752441, 10.315099716186523, 10.318233489990234, 10.31596851348877, 10.313567161560059, 10.31204891204834, 10.31544303894043, 10.314590454101562, 10.310437202453613, 10.316205024719238, 10.318373680114746, 10.313496589660645, 10.313254356384277, 10.309136390686035, 10.31046199798584, 10.314451217651367, 10.310552597045898, 10.308991432189941, 10.314065933227539, 10.313835144042969, 10.312982559204102, 10.314237594604492, 10.309516906738281, 10.313712120056152, 10.315692901611328, 10.31107234954834, 10.315370559692383, 10.31274127960205, 10.30410099029541, 10.313166618347168, 10.305160522460938, 10.30601119

10.30478286743164, 10.290963172912598, 10.301572799682617, 10.292082786560059, 10.29347038269043, 10.293636322021484, 10.294600486755371, 10.296427726745605, 10.293642044067383, 10.295326232910156, 10.293083190917969, 10.294160842895508, 10.298325538635254, 10.29821491241455, 10.298226356506348, 10.29249095916748, 10.291667938232422, 10.297985076904297, 10.2929105758667, 10.295971870422363, 10.29396915435791, 10.291568756103516, 10.295003890991211, 10.295281410217285, 10.296944618225098, 10.291632652282715, 10.299986839294434, 10.293925285339355, 10.298223495483398, 10.296732902526855, 10.295454025268555, 10.29490852355957, 10.296586036682129, 10.290863037109375, 10.29612922668457, 10.295377731323242, 10.295618057250977, 10.296428680419922, 10.291763305664062, 10.29322624206543, 10.293366432189941, 10.295073509216309, 10.29608154296875, 10.294543266296387, 10.293721199035645, 10.294971466064453, 10.293033599853516, 10.295879364013672, 10.294459342956543, 10.294809341430664, 10.29320335

10.283767700195312, 10.283929824829102, 10.283133506774902, 10.283242225646973, 10.282232284545898, 10.283418655395508, 10.285957336425781, 10.283824920654297, 10.288363456726074, 10.282552719116211, 10.28262996673584, 10.280584335327148, 10.282990455627441, 10.287076950073242, 10.281805992126465, 10.283090591430664, 10.284055709838867, 10.283608436584473, 10.281502723693848, 10.284342765808105, 10.281402587890625, 10.280009269714355, 10.289427757263184, 10.285407066345215, 10.282146453857422, 10.283811569213867, 10.28464412689209, 10.276812553405762, 10.280374526977539, 10.285056114196777, 10.285987854003906, 10.280071258544922, 10.284149169921875, 10.284339904785156, 10.28439998626709, 10.281519889831543, 10.285096168518066, 10.286036491394043, 10.288508415222168, 10.282099723815918, 10.284358024597168, 10.284636497497559, 10.283260345458984, 10.286216735839844, 10.283838272094727, 10.280299186706543, 10.283011436462402, 10.283893585205078, 10.284730911254883, 10.281128883361816, 10.

10.271446228027344, 10.269951820373535, 10.2721529006958, 10.274063110351562, 10.273059844970703, 10.272933959960938, 10.272570610046387, 10.271821022033691, 10.273910522460938, 10.269806861877441, 10.274903297424316, 10.267876625061035, 10.281261444091797, 10.268592834472656, 10.274782180786133, 10.272099494934082, 10.270063400268555, 10.277532577514648, 10.274592399597168, 10.271881103515625, 10.276917457580566, 10.273649215698242, 10.274550437927246, 10.276741027832031, 10.268129348754883, 10.274527549743652, 10.2694091796875, 10.271145820617676, 10.271862983703613, 10.270753860473633, 10.270052909851074, 10.269682884216309, 10.268794059753418, 10.275880813598633, 10.272594451904297, 10.270610809326172, 10.272688865661621, 10.27167797088623, 10.271001815795898, 10.272456169128418, 10.271580696105957, 10.274897575378418, 10.274370193481445, 10.271479606628418, 10.267799377441406, 10.272514343261719, 10.27328109741211, 10.271124839782715, 10.269509315490723, 10.27608871459961, 10.2668

10.262619972229004, 10.260927200317383, 10.260445594787598, 10.26146411895752, 10.261025428771973, 10.25885009765625, 10.265101432800293, 10.259918212890625, 10.265250205993652, 10.264763832092285, 10.26237678527832, 10.257946968078613, 10.262846946716309, 10.269768714904785, 10.266072273254395, 10.262261390686035, 10.25922966003418, 10.263711929321289, 10.265524864196777, 10.264124870300293, 10.264294624328613, 10.262929916381836, 10.262636184692383, 10.262899398803711, 10.260534286499023, 10.262969970703125, 10.26065731048584, 10.26224136352539, 10.25810718536377, 10.266438484191895, 10.260870933532715, 10.259214401245117, 10.259902000427246, 10.262953758239746, 10.26297378540039, 10.267762184143066, 10.261541366577148, 10.264617919921875, 10.259154319763184, 10.256677627563477, 10.267081260681152, 10.262136459350586, 10.26254653930664, 10.260024070739746, 10.265863418579102, 10.265914916992188, 10.260401725769043, 10.262141227722168, 10.256979942321777, 10.26012897491455, 10.2635145

10.254143714904785, 10.251287460327148, 10.253124237060547, 10.249861717224121, 10.256481170654297, 10.251437187194824, 10.253077507019043, 10.252678871154785, 10.250893592834473, 10.253005027770996, 10.252678871154785, 10.251511573791504, 10.25172233581543, 10.255844116210938, 10.258673667907715, 10.253945350646973, 10.252327919006348, 10.255067825317383, 10.250800132751465, 10.25133991241455, 10.255926132202148, 10.252761840820312, 10.248980522155762, 10.252352714538574, 10.252717018127441, 10.253861427307129, 10.253982543945312, 10.250426292419434, 10.250703811645508, 10.255980491638184, 10.254405975341797, 10.25313949584961, 10.251282691955566, 10.25688648223877, 10.25553035736084, 10.257163047790527, 10.251777648925781, 10.250924110412598, 10.252472877502441, 10.253265380859375, 10.254151344299316, 10.256227493286133, 10.254766464233398, 10.247740745544434, 10.250591278076172, 10.250358581542969, 10.255769729614258, 10.253072738647461, 10.25704288482666, 10.255973815917969, 10.254

10.242547035217285, 10.240212440490723, 10.245353698730469, 10.245319366455078, 10.245858192443848, 10.24730396270752, 10.247559547424316, 10.244630813598633, 10.241556167602539, 10.240839004516602, 10.2435302734375, 10.244953155517578, 10.244681358337402, 10.244915962219238, 10.246358871459961, 10.247665405273438, 10.246831893920898, 10.244571685791016, 10.243532180786133, 10.244606971740723, 10.24225902557373, 10.242655754089355, 10.247467994689941, 10.243045806884766, 10.246733665466309, 10.243889808654785, 10.24339485168457, 10.246417045593262, 10.246219635009766, 10.248202323913574, 10.243383407592773, 10.241315841674805, 10.245920181274414, 10.240355491638184, 10.241012573242188, 10.2462158203125, 10.243805885314941, 10.246515274047852, 10.241300582885742, 10.245216369628906, 10.243757247924805, 10.244810104370117, 10.24227523803711, 10.244194030761719, 10.239502906799316, 10.24473762512207, 10.245464324951172, 10.246127128601074, 10.242959976196289, 10.247672080993652, 10.241252

10.234946250915527, 10.236130714416504, 10.237300872802734, 10.237253189086914, 10.235021591186523, 10.242796897888184, 10.236589431762695, 10.235417366027832, 10.24142074584961, 10.235434532165527, 10.241450309753418, 10.23033332824707, 10.235724449157715, 10.23586368560791, 10.234959602355957, 10.235771179199219, 10.237181663513184, 10.238402366638184, 10.237342834472656, 10.23534870147705, 10.23430347442627, 10.23614501953125, 10.22968864440918, 10.236891746520996, 10.242773056030273, 10.23491382598877, 10.231734275817871, 10.235616683959961, 10.235674858093262, 10.23531723022461, 10.232892990112305, 10.23216438293457, 10.238472938537598, 10.24074935913086, 10.233379364013672, 10.238903999328613, 10.232672691345215, 10.234469413757324, 10.235087394714355, 10.239165306091309, 10.233173370361328, 10.237184524536133, 10.23457145690918, 10.239056587219238, 10.23844051361084, 10.235457420349121, 10.231293678283691, 10.231945037841797, 10.235206604003906, 10.236137390136719, 10.2349634170

10.224761962890625, 10.22740364074707, 10.229272842407227, 10.224589347839355, 10.232876777648926, 10.225483894348145, 10.23156452178955, 10.228660583496094, 10.228642463684082, 10.229771614074707, 10.22640609741211, 10.229018211364746, 10.229586601257324, 10.22752571105957, 10.223700523376465, 10.231134414672852, 10.23210334777832, 10.229114532470703, 10.230170249938965, 10.223835945129395, 10.229065895080566, 10.229700088500977, 10.227025032043457, 10.228543281555176, 10.226958274841309, 10.228798866271973, 10.226758003234863, 10.224581718444824, 10.228684425354004, 10.222164154052734, 10.232963562011719, 10.229242324829102, 10.231094360351562, 10.228596687316895, 10.225910186767578, 10.228336334228516, 10.232486724853516, 10.228693008422852, 10.224607467651367, 10.22468090057373, 10.225409507751465, 10.22893238067627, 10.228623390197754, 10.230424880981445, 10.229583740234375, 10.22906494140625, 10.232629776000977, 10.229671478271484, 10.227181434631348, 10.224223136901855, 10.22663

10.228909492492676, 10.223947525024414, 10.224092483520508, 10.224543571472168, 10.223194122314453, 10.223017692565918, 10.224357604980469, 10.224209785461426, 10.222232818603516, 10.222922325134277, 10.220561981201172, 10.229923248291016, 10.221733093261719, 10.22302532196045, 10.222084045410156, 10.223438262939453, 10.228765487670898, 10.221206665039062, 10.22155475616455, 10.223406791687012, 10.22494888305664, 10.22741985321045, 10.22127628326416, 10.220389366149902, 10.223527908325195, 10.22177505493164, 10.226601600646973, 10.219860076904297, 10.221745491027832, 10.226664543151855, 10.219436645507812, 10.22415542602539, 10.22732925415039, 10.219372749328613, 10.218865394592285, 10.225308418273926, 10.223690032958984, 10.22006607055664, 10.218730926513672, 10.221365928649902, 10.220911979675293, 10.222477912902832, 10.217391014099121, 10.221051216125488, 10.225433349609375, 10.22262954711914, 10.219839096069336, 10.23325252532959, 10.220973014831543, 10.21810245513916, 10.225475311

10.21730899810791, 10.216856956481934, 10.218402862548828, 10.217668533325195, 10.21939754486084, 10.21433162689209, 10.220257759094238, 10.217941284179688, 10.218643188476562, 10.217313766479492, 10.220815658569336, 10.215497970581055, 10.215494155883789, 10.222734451293945, 10.220854759216309, 10.218816757202148, 10.216733932495117, 10.215934753417969, 10.219422340393066, 10.216740608215332, 10.217855453491211, 10.22150707244873, 10.221894264221191, 10.216927528381348, 10.218863487243652, 10.21996784210205, 10.215937614440918, 10.2210111618042, 10.21967601776123, 10.215180397033691, 10.218958854675293, 10.218128204345703, 10.217766761779785, 10.219337463378906, 10.217061042785645, 10.217667579650879, 10.221380233764648, 10.215574264526367, 10.216870307922363, 10.216899871826172, 10.213994979858398, 10.215292930603027, 10.21312427520752, 10.22116756439209, 10.218613624572754, 10.219717025756836, 10.219171524047852, 10.220040321350098, 10.219868659973145, 10.218411445617676, 10.2163238

10.217594146728516, 10.210453987121582, 10.209402084350586, 10.21205997467041, 10.214550018310547, 10.210954666137695, 10.213065147399902, 10.215622901916504, 10.211838722229004, 10.213516235351562, 10.217329978942871, 10.217350959777832, 10.214166641235352, 10.212203025817871, 10.211584091186523, 10.209515571594238, 10.212644577026367, 10.210111618041992, 10.214520454406738, 10.21146297454834, 10.216190338134766, 10.214681625366211, 10.213846206665039, 10.214966773986816, 10.214344024658203, 10.216261863708496, 10.216092109680176, 10.215261459350586, 10.210403442382812, 10.211742401123047, 10.21306037902832, 10.217960357666016, 10.217479705810547, 10.215717315673828, 10.210357666015625, 10.208675384521484, 10.219587326049805, 10.213189125061035, 10.214136123657227, 10.210221290588379, 10.2109375, 10.215510368347168, 10.207810401916504, 10.214205741882324, 10.215084075927734, 10.210325241088867, 10.20970344543457, 10.214872360229492, 10.210503578186035, 10.216033935546875, 10.216620445

10.207953453063965, 10.207883834838867, 10.207527160644531, 10.208725929260254, 10.207771301269531, 10.206258773803711, 10.208145141601562, 10.208159446716309, 10.207987785339355, 10.209508895874023, 10.20935344696045, 10.206942558288574, 10.206900596618652, 10.209486961364746, 10.205013275146484, 10.20931339263916, 10.207016944885254, 10.210288047790527, 10.207576751708984, 10.210003852844238, 10.208322525024414, 10.20926284790039, 10.206392288208008, 10.207622528076172, 10.207440376281738, 10.213780403137207, 10.20853328704834, 10.20726203918457, 10.207549095153809, 10.204448699951172, 10.20775318145752, 10.21550178527832, 10.209959983825684, 10.204283714294434, 10.20522689819336, 10.209728240966797, 10.2131986618042, 10.206775665283203, 10.207565307617188, 10.210044860839844, 10.21066665649414, 10.206098556518555, 10.209321022033691, 10.209357261657715, 10.21132755279541, 10.205766677856445, 10.207047462463379, 10.205667495727539, 10.205832481384277, 10.212810516357422, 10.212116241

10.205192565917969, 10.207000732421875, 10.202558517456055, 10.208416938781738, 10.205802917480469, 10.201812744140625, 10.202725410461426, 10.198649406433105, 10.203556060791016, 10.20697021484375, 10.203592300415039, 10.204753875732422, 10.2039213180542, 10.204141616821289, 10.204055786132812, 10.204879760742188, 10.205310821533203, 10.20627212524414, 10.207159042358398, 10.20808219909668, 10.200845718383789, 10.207062721252441, 10.203288078308105, 10.204103469848633, 10.20678424835205, 10.202285766601562, 10.20561408996582, 10.202993392944336, 10.202098846435547, 10.2027587890625, 10.201300621032715, 10.204837799072266, 10.201014518737793, 10.202437400817871, 10.208267211914062, 10.20390796661377, 10.203934669494629, 10.20421028137207, 10.202451705932617, 10.203739166259766, 10.206840515136719, 10.206335067749023, 10.209610939025879, 10.203271865844727, 10.197952270507812, 10.197551727294922, 10.203694343566895, 10.203010559082031, 10.204320907592773, 10.19848918914795, 10.199051856